libs/wingc/src/type_check.rs

```diff
@@ -4778,6 +4769,38 @@ new cloud.Function(@inflight("./handler.ts"), lifts: { bucket: ["put"] });
 		self.ctx.pop_class();
 	}
 
+	fn type_check_invalid_expr_before_super(&mut self, expr: &Expr) -> bool {
+		match &expr.kind {
+			ExprKind::Call { callee, .. } => {
+				// callee is SuperCall return true
+				if matches!(callee, CalleeKind::SuperCall(_)) {
+					return true;
+				}
+				// if callee is an expression checks if has an instante member
+				return match callee {
+					CalleeKind::Expr(e) => {
+						matches!(e.kind, ast::ExprKind::Reference(Reference::InstanceMember { .. }))
+					}
+					_ => false,
+				};
+			}
+			_ => false,
+		}
+	}
+	fn type_check_invalid_stmt_before_super(&mut self, stmt: &Stmt) -> bool {
```

Comment:
The logic here seems to simplistic, or to narrow and only handles certain statements. I'd expect this to be a function that creates some visitor object to recursively run over all inner expression of the statement and search for a reference to `this` or ` CalleeKind::SuperCall` in an `ExprKind::Call`.
This needs to be a recursive visitor because there might be some inner scope (like an `if` block) that references `this` or calls `super.something()`.

```rust
//psuedo code
let mut check = CheckValidBeforeSuperVisitor::new();
check.visit_stmt(stmt);
return check.stmt_valid;
```
The implementation of the `Visitor` trait needs to internally do:
* implement `visit_expr` and set `stmt_valid` to false if `matches!(expr.kind, ExprKind::Call { calee:  CalleeKind::SuperCall(_), .. }`
* implement `visit_reference` and set `stmt_valid` to false if the reference matches `Reference::Identifier(Symbol {name: "this", ..})`
* Make sure not to dive into class definitions, so inner references to `this` or super method calls remain valid. This can be done by impelmenting `visit_class` in our visitor and just retuning without looking into the inner parts of the class.

Let me know if this makes sense or if something isn't clear...